tests/e2e/utils/publishPost.js

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * WordPress dependencies
+ */
+import {
+  arePrePublishChecksEnabled,
+  publishPostWithPrePublishChecksDisabled,
+  publishPost as _publishPost,
+} from '@wordpress/e2e-test-utils';
+
+/**
+ * Custom helper function to publish posts in Gutenberg.
+ *
+ * Avoid using disablePrePublishChecks.
+ *
+ * The main reason for this is that it calls toggleScreenOption,
+ * which itself is rather flaky and sometimes
+ * doesn't find the requested element, causing JS errors.
+ *
+ * @see https://github.com/WordPress/gutenberg/blob/f96bd6b6fdfca02d47edce8c2f0d68dccf3a37be/packages/e2e-test-utils/src/toggle-screen-option.js#L20-L23
+ *
+ * @return {Promise<string>} The post's permalink.
+ */
+async function publishPost() {
+  const prePublishChecksEnabled = await arePrePublishChecksEnabled();
+
+  if (prePublishChecksEnabled) {
+    await _publishPost();
+  } else {
+    await publishPostWithPrePublishChecksDisabled();
+  }
+
+  // Wait until the selector returns a truthy value.
+  await page.waitForFunction(() =>
+    wp.data.select('core/editor').getPermalink()
+  );
+
+  return page.evaluate(() => wp.data.select('core/editor').getPermalink());
```

Comment:
Why is this run twice?

Reply:
The `// Wait until the selector returns a truthy value.` comment above explains it.

`wp.data.select('core/editor').getPermalink()` does not always return a permalink string right away. For example because in the background some data is still being fetched. Then it returns `null` instead.

The `waitForFunction` call above waits until a truthy value (i.e. not `null`) is returned. But it returns a puppeteer JSHandle object, not the actual value.

So with `page.evaluate` we get the actual value after we know it's truthy.